File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int pivot = -1;

        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                pivot = i;
                break;
            }
        }

        if (pivot == -1) {
            reverse(nums, 0);
            return;
        }

        // int next = -1;
        for (int i = nums.length - 1; i >= 0; i--) {
            if (nums[i] > nums[pivot]) {
                // next = i;
                swap(nums, i, pivot);
                break;
            }
        }

        reverse(nums, pivot + 1);
    }

    private void swap (int[] nums, int start, int end) {
        int temp = nums[start];
        nums[start] = nums[end];
        nums[end] = temp;
    }

    private void reverse (int[] nums, int start) {
        int end = nums.length - 1;

        while (start < end) {
            swap (nums, start, end);
            start++;
            end--;
        }
    }
}

File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int toswap = -1;
        for (int i = n - 1; i > 0; i--) {
            if (nums[i] > nums[i - 1]) {
                toswap = i - 1;
                break;
            }
        }
        if (toswap == -1) {
            reverse(nums.begin(), nums.end());
            return;
        }
        for (int i = n - 1; i > toswap; i--) {
            if (nums[i] > nums[toswap]) {
                swap(nums[i], nums[toswap]);
                break;
            }
        }
        reverse(nums.begin() + toswap + 1, nums.end());
    }
};

